File: 1870-minimum-speed-to-arrive-on-time/1870-minimum-speed-to-arrive-on-time.cpp
class Solution {
public:
    int minSpeedOnTime(vector<int>& dist, double hour) {
        if (dist.size() > ceil(hour)) return -1;
        
        int left = 1, right = pow(10, 7);
        while (left <= right){
            int mid = left + (right - left) / 2;
            double time = 0.0;
            
            for (double d : dist){
                time = ceil(time);
                time += d / mid;
            }
            
            if (time <= hour) right = mid - 1;
            else left = mid + 1;
        }
        
        return left;
    }
};

File: 1870-minimum-speed-to-arrive-on-time/1870-minimum-speed-to-arrive-on-time.py
class Solution:
    def minSpeedOnTime(self, dist: List[int], hour: float) -> int:
        if len(dist) > ceil(hour):
            return -1
        
        left, right = 1, 10 ** 7
        
        while left <= right:
            mid = (left + right) // 2
            time = 0
            
            for i in range(len(dist)):
                time = ceil(time)
                time += dist[i] / mid
            
            if time <= hour:
                right = mid - 1
            else:
                left = mid + 1
        
        return left
        